File: app/src/main/java/com/codetest/app/auth/AuthenticatedUserArgumentResolver.java
package com.codetest.app.auth;

import com.codetest.auth.AuthResult;
import com.codetest.auth.AuthService;
import com.codetest.auth.UserDetails;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.core.MethodParameter;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;
import org.springframework.web.server.ResponseStatusException;

public class AuthenticatedUserArgumentResolver implements HandlerMethodArgumentResolver {

    private static final String BEARER_PREFIX = "Bearer ";

    private final AuthService authService;

    public AuthenticatedUserArgumentResolver(AuthService authService) {
        this.authService = authService;
    }

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(AuthenticatedUser.class)
                && parameter.getParameterType().equals(UserDetails.class);
    }

    @Override
    public Object resolveArgument(MethodParameter parameter,
            ModelAndViewContainer mavContainer,
            NativeWebRequest webRequest,
            WebDataBinderFactory binderFactory) {
        HttpServletRequest request = webRequest.getNativeRequest(HttpServletRequest.class);
        return resolveUserDetails(request);
    }

    private UserDetails resolveUserDetails(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");

        if (authHeader == null || !authHeader.startsWith(BEARER_PREFIX)) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED);
        }

        String token = authHeader.substring(BEARER_PREFIX.length());

        AuthResult result = authService.authenticate(token);

        if (result instanceof AuthResult.Authenticated auth) {
            return auth.user();
        }

        // Fail with 401 as expected
        throw new ResponseStatusException(HttpStatus.UNAUTHORIZED);
    }
}

File: app/src/main/scala/com/codetest/app/auth/AuthenticatedUserArgumentResolver.scala
package com.codetest.app.auth

import com.codetest.auth.{AuthService, UserDetails}
import jakarta.servlet.http.HttpServletRequest
import org.springframework.core.MethodParameter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.{HandlerMethodArgumentResolver, ModelAndViewContainer}

class AuthenticatedUserArgumentResolver(authService: AuthService) extends HandlerMethodArgumentResolver {

  private val BearerPrefix = "Bearer "

  override def supportsParameter(parameter: MethodParameter): Boolean =
    parameter.hasParameterAnnotation(classOf[AuthenticatedUser]) &&
      parameter.getParameterType == classOf[UserDetails]

  override def resolveArgument(
      parameter: MethodParameter,
      mavContainer: ModelAndViewContainer,
      webRequest: NativeWebRequest,
      binderFactory: WebDataBinderFactory
  ): AnyRef = {
    val request = webRequest.getNativeRequest(classOf[HttpServletRequest])
    resolveUserDetails(request)
  }

  private def resolveUserDetails(request: HttpServletRequest): UserDetails = {
    /* TODO (Candidate)
     *  Extract the token using `request.getHeader("Authorization")`
     *  and use `authService.authenticate(String jwt)` to get the user if the token is valid,
     *  or return a 401 Status if not valid, or malformed etc..
     *  returning a `UserDetails` object from resolveArgument will automagically inject it into the Controller method call.
     */
    throw new UnsupportedOperationException("Implement Me")
  }
}
